### ISSUE TITLE
this branch never matches

### ISSUE THREAD

**Issue by skogsbaer** [etylizer/etylizer#171]
```erlang
-module(test).

-compile(export_all).
-compile(nowarn_export_all).

-spec range0(#{ any() => V }) -> [V].
range0(Map) ->
    lists:map(fun({_, V}) -> V end, maps:to_list(Map))
```

fails with

```
[NOTE 2024-11-15 10:25:22.137 typing_check.erl:29] Checking failed: test.erl:8:18: Type error: this branch never matches
%    8|     lists:map(fun({_, V}) -> V end, maps:to_list(Map)).
%     |                  ^
test.erl:8:18: Type error: this branch never matches
%    8|     lists:map(fun({_, V}) -> V end, maps:to_list(Map)).
%     |
```

Interestingly, if we split the function in two, type checking succeeds:

```
-spec range2(#{ any() => V }) -> [V].
range2(Map) ->
    range1(maps:to_list(Map)).

-spec range1([{any(), V}]) -> [V].
range1(L) ->
    lists:map(fun({_, V}) -> V end, L).

```

**Comment by skogsbaer** [comment]
Smaller example that still fails:

```erlang
-spec to_list(MapOrIterator) -> [{Key, Value}] when MapOrIterator :: #{Key => Value}.
to_list(X) -> to_list(X).

-spec range(#{ any() => V }) -> [V].
range(Map) ->
    lists:map(fun({_, V}) -> V end, to_list2(Map)).
```

Expanding `MapOrIterator` "solves" the problem:

```erlang
-spec to_list(#{Key => Value}) -> [{Key, Value}].
to_list(X) -> to_list(X).

-spec range(#{ any() => V }) -> [V].
range(Map) ->
    lists:map(fun({_, V}) -> V end, to_list2(Map)).
```

Hence, I think this bug should be fixed once #155 is fixed.

**Comment by skogsbaer** [comment]
Duplicate of #155
